src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp

```diff
@@ -0,0 +1,233 @@
+             << " stored quaternions. There must be at least the same number "
+                "of stored quaternions as omegas otherwise we cannot solve the "
+                "ode and populate the missing quaternions.");
+
+  // If they are already the same size, we don't have to update anything so exit
+  // now
+  if (omega_deriv_info.size() == stored_quaternions_and_times_.size()) {
+    return;
+  }
+
+  // Will store the solution of the ode
+  std::array<DataVector, 1> quaternion_to_integrate;
+  for (size_t i = stored_quaternions_and_times_.size();
+       i < omega_deriv_info.size(); i++) {
+    // Final time, initial time, and quaternion
+    const double t = omega_deriv_info[i].time;
+    const double t0 = omega_deriv_info[i - 1].time;
+    quaternion_to_integrate =
+        stored_quaternions_and_times_[i - 1].stored_quantities;
+
+    solve_quaternion_ode(&quaternion_to_integrate, t, t0);
+
+    // normalize quaternion
+    boost::math::quaternion<double> quat =
+        datavector_to_quaternion(quaternion_to_integrate[0]);
+    normalize_quaternion(make_not_null(&quat));
+
+    stored_quaternions_and_times_.emplace_back(
+        //  FunctionOfTimeHelpers::StoredInfo<1, false>{
+        t, std::array<DataVector, 1>{quaternion_to_datavector(quat)});
+  }
+
+  ASSERT(omega_deriv_info.size() == stored_quaternions_and_times_.size(),
+         "The number of stored omegas must be the same as the number of stored "
+         "quaternions after updating the missing quaternions. Now there are "
+             << omega_deriv_info.size() << " stored omegas and "
+             << stored_quaternions_and_times_.size() << " stored quaternions.");
+}
+
+template <size_t MaxDeriv>
+void QuaternionFunctionOfTime<MaxDeriv>::solve_quaternion_ode(
+    const gsl::not_null<std::array<DataVector, 1>*> quaternion_to_integrate,
+    const double t, const double t0) const noexcept {
+  // lambda that stores the internals of the ode
+  const auto quaternion_ode_system =
+      [this](const std::array<DataVector, 1>& state,
+             std::array<DataVector, 1>& dt_state, const double time) noexcept {
+        const std::array<DataVector, 1> omega = omega_f_of_t_.func(time);
+        dt_state[0] = DataVector{
+            {-0.5 * (state[0][1] * omega[0][0] + state[0][2] * omega[0][1] +
+                     state[0][3] * omega[0][2]),
+             0.5 * (state[0][0] * omega[0][0] + state[0][2] * omega[0][2] -
+                    state[0][3] * omega[0][1]),
+             0.5 * (state[0][0] * omega[0][1] + state[0][3] * omega[0][0] -
+                    state[0][1] * omega[0][2]),
+             0.5 * (state[0][0] * omega[0][2] + state[0][1] * omega[0][1] -
+                    state[0][2] * omega[0][0])}};
+      };
+
+  // Dense stepper
+  boost::numeric::odeint::dense_output_runge_kutta<
+      boost::numeric::odeint::controlled_runge_kutta<
+          boost::numeric::odeint::runge_kutta_dopri5<
+              std::array<DataVector, 1>>>>
```

Comment:
If there's no conversion going on here, make this `auto`.  Seeing the full type isn't informative.